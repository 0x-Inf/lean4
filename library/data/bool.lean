-- Copyright (c) 2014 Microsoft Corporation. All rights reserved.
-- Released under Apache 2.0 license as described in the file LICENSE.
-- Author: Leonardo de Moura
import general_notation
import logic.connectives logic.decidable logic.inhabited

open eq eq.ops decidable

inductive bool : Type :=
  ff : bool,
  tt : bool
namespace bool
  protected definition cases_on {p : bool → Prop} (b : bool) (H₁ : p ff) (H₂ : p tt) : p b :=
  rec H₁ H₂ b

  definition cond {A : Type} (b : bool) (t e : A) :=
  rec_on b e t

  theorem dichotomy (b : bool) : b = ff ∨ b = tt :=
  cases_on b (or.inl rfl) (or.inr rfl)

  theorem cond.ff {A : Type} (t e : A) : cond ff t e = e :=
  rfl

  theorem cond.tt {A : Type} (t e : A) : cond tt t e = t :=
  rfl

  theorem ff_ne_tt : ¬ ff = tt :=
  assume H : ff = tt, absurd
    (calc true  = cond tt true false : !cond.tt⁻¹
            ... = cond ff true false : {H⁻¹}
            ... = false              : !cond.ff)
    true_ne_false

  definition bor (a b : bool) :=
  rec_on a (rec_on b ff tt) tt

  theorem bor.tt_left (a : bool) : bor tt a = tt :=
  rfl

  notation a || b := bor a b

  theorem bor.tt_right (a : bool) : a || tt = tt :=
  cases_on a rfl rfl

  theorem bor.ff_left (a : bool) : ff || a = a :=
  cases_on a rfl rfl

  theorem bor.ff_right (a : bool) : a || ff = a :=
  cases_on a rfl rfl

  theorem bor.id (a : bool) : a || a = a :=
  cases_on a rfl rfl

  theorem bor.comm (a b : bool) : a || b = b || a :=
  cases_on a
    (cases_on b rfl rfl)
    (cases_on b rfl rfl)

  theorem bor.assoc (a b c : bool) : (a || b) || c = a || (b || c) :=
  cases_on a
    (calc (ff || b) || c = b || c         : {!bor.ff_left}
                   ...   = ff || (b || c) : !bor.ff_left⁻¹)
    (calc (tt || b) || c = tt || c        : {!bor.tt_left}
                   ...   = tt             : !bor.tt_left
                   ...   = tt || (b || c) : !bor.tt_left⁻¹)

  theorem bor.to_or {a b : bool} : a || b = tt → a = tt ∨ b = tt :=
  rec_on a
    (assume H : ff || b = tt,
      have Hb : b = tt, from !bor.ff_left ▸ H,
      or.inr Hb)
    (assume H, or.inl rfl)

  definition band (a b : bool) :=
  rec_on a ff (rec_on b ff tt)

  notation a && b := band a b

  theorem band.ff_left (a : bool) : ff && a = ff :=
  rfl

  theorem band.tt_left (a : bool) : tt && a = a :=
  cases_on a rfl rfl

  theorem band.ff_right (a : bool) : a && ff = ff :=
  cases_on a rfl rfl

  theorem band.tt_right (a : bool) : a && tt = a :=
  cases_on a rfl rfl

  theorem band.id (a : bool) : a && a = a :=
  cases_on a rfl rfl

  theorem band.comm (a b : bool) : a && b = b && a :=
  cases_on a
    (cases_on b rfl rfl)
    (cases_on b rfl rfl)

  theorem band.assoc (a b c : bool) : (a && b) && c = a && (b && c) :=
  cases_on a
    (calc (ff && b) && c = ff && c        : {!band.ff_left}
                    ...  = ff             : !band.ff_left
                    ...  = ff && (b && c) : !band.ff_left⁻¹)
    (calc (tt && b) && c = b && c         : {!band.tt_left}
                    ...  = tt && (b && c) : !band.tt_left⁻¹)

  theorem band.eq_tt_elim_left {a b : bool} (H : a && b = tt) : a = tt :=
  or.elim (dichotomy a)
    (assume H0 : a = ff,
      absurd
        (calc ff = ff && b : !band.ff_left⁻¹
             ... = a && b  : {H0⁻¹}
             ... = tt      : H)
        ff_ne_tt)
    (assume H1 : a = tt, H1)

  theorem band.eq_tt_elim_right {a b : bool} (H : a && b = tt) : b = tt :=
  band.eq_tt_elim_left (!band.comm ⬝ H)

  definition bnot (a : bool) :=
  rec_on a tt ff

  theorem bnot.bnot (a : bool) : bnot (bnot a) = a :=
  cases_on a rfl rfl

  theorem bnot.false : bnot ff = tt :=
  rfl

  theorem bnot.true  : bnot tt = ff :=
  rfl

  protected definition is_inhabited [instance] : inhabited bool :=
  inhabited.mk ff

  protected definition has_decidable_eq [instance] : decidable_eq bool :=
  take a b : bool,
    rec_on a
      (rec_on b (inl rfl) (inr ff_ne_tt))
      (rec_on b (inr (ne.symm ff_ne_tt)) (inl rfl))
end bool
