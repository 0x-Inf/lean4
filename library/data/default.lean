/-
Copyright (c) 2014 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Jeremy Avigad
-/

import .empty .bool .num .nat .list
--  .set
-- import .string .int .rat .fintype .prod .sum .sigma .option .list .finset .set .stream
-- import .fin .real .complex
