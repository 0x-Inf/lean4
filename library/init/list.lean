/-
Copyright (c) 2016 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Leonardo de Moura
-/
prelude
import init.logic
open decidable list

protected definition list.is_inhabited [instance] (A : Type) : inhabited (list A) :=
inhabited.mk list.nil

definition list.has_decidable_eq [instance] {A : Type} [H : decidable_eq A] (l₁ : list A) :  ∀ l₂ : list A, decidable (l₁ = l₂) :=
list.rec_on l₁
  (λ l₂, list.cases_on l₂
    (tt rfl)
    (λ b l₂, ff (λ H, list.no_confusion H)))
  (λ a l₁ ih l₂, list.cases_on l₂
    (ff (λ H, list.no_confusion H))
    (λ b l₂,
       decidable.cases_on (H a b)
        (λ Hnab : a ≠ b, ff (λ H, list.no_confusion H (λ Hab Hl₁l₂, absurd Hab Hnab)))
        (λ Hab : a = b,
           decidable.cases_on (ih l₂)
             (λ Hne : l₁ ≠ l₂, ff (λ H, list.no_confusion H (λ Hab Hl₁l₂, absurd Hl₁l₂ Hne)))
             (λ He  : l₁ = l₂, tt (congr (congr_arg cons Hab) He)))))

notation h :: t  := cons h t
notation `[` l:(foldr `, ` (h t, cons h t) nil `]`) := l

namespace list

variable {A : Type}

definition append : list A → list A → list A
| []       l := l
| (h :: s) t := h :: (append s t)

notation l₁ ++ l₂ := append l₁ l₂

definition length : list A → nat
| []       := 0
| (a :: l) := length l + 1

end list
