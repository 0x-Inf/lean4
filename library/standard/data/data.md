data
====

Various datatypes.

Basic types:

* [empty](empty.lean) : the empty type
* [unit](unit.lean) : the singleton type
* [bool](bool/bool.md) : the boolean values
* [num](num.lean) : generic numerals
* [string](string.lean) : ascii strings
* [nat](nat/nat.md) : the natural numbers
* [int](int/int.md) : the integers

Constructors:

* [option](option.lean)
* [pair](pair.lean)
* [list](list/list.md)
* [set](set.lean)