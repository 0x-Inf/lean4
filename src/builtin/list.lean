-- Copyright (c) 2014 Microsoft Corporation. All rights reserved.
-- Released under Apache 2.0 license as described in the file LICENSE.
-- Author: Leonardo de Moura
import num subtype optional macros tactic
using num
using subtype

namespace list
definition none {A : (Type U)} : optional A
:= optional::@none A

definition some {A : (Type U)} (a : A) : optional A
:= optional::some a

definition list_rep (A : (Type U))
:= (num → optional A) # num

definition list_pred (A : (Type U))
:= λ p : list_rep A, ∀ i, i < (proj2 p) ↔ (proj1 p) i ≠ (@none A)

definition list (A : (Type U))
:= subtype (list_rep A) (list_pred A)

definition len {A : (Type U)} (l : list A) : num
:= proj2 (rep l)

definition fn {A : (Type U)} (l : list A) : num → optional A
:= proj1 (rep l)

theorem ext {A : (Type U)} {l1 l2 : list A} (H1 : len l1 = len l2) (H2 : fn l1 = fn l2) : l1 = l2
:= have Heq : rep l1 = rep l2,
     from pairext _ _ H2 (to_heq H1),
   rep_inj Heq

definition nil_rep (A : (Type U)) : list_rep A
:= (pair (λ n : num, @none A) zero : list_rep A)

theorem nil_pred (A : (Type U)) : list_pred A (nil_rep A)
:= take i : num,
   let nil := nil_rep A
   in iff_intro
     (assume Hl : i < (proj2 nil),
      have H1 : i < zero,
        from Hl,
      absurd_elim ((proj1 nil) i ≠ (@none A)) H1 (not_lt_zero i))
     (assume Hr : (proj1 nil) i ≠ (@none A),
      have H1 : (@none A) ≠ (@none A),
        from Hr,
      false_elim (i < (proj2 nil)) (a_neq_a_elim H1))

theorem inhab (A : (Type U)) : inhabited (list A)
:= subtype_inhabited (exists_intro (nil_rep A) (nil_pred A))

definition nil {A : (Type U)} : list A
:= abst (nil_rep A) (list::inhab A)

theorem len_nil {A : (Type U)} : len (@nil A) = zero
:= have H1 : rep (@nil A) = (pair (λ n : num, @none A) zero : list_rep A),
     from rep_abst (list::inhab A) (nil_rep A) (nil_pred A),
   have H2 : len (@nil A) = proj2 (rep (@nil A)),
     from refl (len (@nil A)),
   have H3 : proj2 (rep (@nil A)) = zero,
     from proj2_congr H1,
   trans H2 H3

definition cons_rep {A : (Type U)} (h : A) (t : list A) : list_rep A
:= pair (λ n, if n = (len t) then (some h) else (fn t) n) (succ (len t))

theorem cons_rep_fn_at {A : (Type U)} (h : A) (t : list A) (i : num)
                       : (proj1 (cons_rep h t)) i = (if i = len t then some h else (fn t) i)
:= have Heq : proj1 (cons_rep h t) = λ n, if n = len t then some h else (fn t) n,
     from refl (proj1 (cons_rep h t)),
   congr1 Heq i

definition cons_pred {A : (Type U)} (h : A) (t : list A) : list_pred A (cons_rep h t)
:= take i : num,
   let  c   := cons_rep h t in
   have Hci : (proj1 c) i = (if i = (len t) then (some h) else (fn t) i),
     from cons_rep_fn_at h t i,
   have Ht  : ∀ i, i < (len t) ↔ (fn t) i ≠ (@none A),
     from P_rep t,
   iff_intro
     (assume Hl : i < (succ (len t)),
      or_elim (em (i = len t))
         (assume Heq : i = len t,
           calc (proj1 c) i = (if i = (len t) then (some h) else (fn t) i) : Hci
                        ... = some h                                       : by simp
                        ... ≠ @none A                                      : optional::distinct h)
         (assume Hne : i ≠ len t,
           have Hlt : i < len t,
             from lt_succ_ne_to_lt Hl Hne,
           calc (proj1 c) i = (if i = (len t) then (some h) else (fn t) i) : Hci
                        ... = (fn t) i                                     : by simp
                        ... ≠ @none A                                      : (Ht i) ◂ Hlt))
     (assume Hr : (proj1 c) i ≠ (@none A),
      or_elim (em (i = len t))
         (assume Heq : i = len t,
           subst (n_lt_succ_n (len t)) (symm Heq))
         (assume Hne : i ≠ len t,
           have Hne2 : (fn t) i ≠ (@none A),
             from calc (fn t) i = (if i = (len t) then (some h) else (fn t) i) : by simp
                            ... = (proj1 c) i                                  : symm Hci
                            ... ≠ (@none A)                                    : Hr,
           have Hlt : i < len t,
             from (symm (Ht i)) ◂ Hne2,
           show i < succ (len t),
             from lt_to_lt_succ Hlt))

definition cons {A : (Type U)} (h : A) (t : list A) : list A
:= abst (cons_rep h t) (list::inhab A)

theorem cons_fn_at {A : (Type U)} (h : A) (t : list A) (i : num)
                       : (fn (cons h t)) i = (if i = len t then some h else (fn t) i)
:= have H1 : rep (cons h t) = (cons_rep h t) ,
     from rep_abst (list::inhab A) (cons_rep h t) (cons_pred h t),
   have H2 : fn (cons h t) = proj1 (cons_rep h t),
     from proj1_congr H1,
   have H3 : fn (cons h t) i = (proj1 (cons_rep h t)) i,
     from congr1 H2 i,
   have H4 : (proj1 (cons_rep h t)) i = (if i = len t then some h else (fn t) i),
     from cons_rep_fn_at h t i,
   trans H3 H4

theorem len_cons {A : (Type U)} (h : A) (t : list A) : len (cons h t) = succ (len t)
:= have H1 : rep (cons h t) = cons_rep h t,
     from rep_abst (list::inhab A) (cons_rep h t) (cons_pred h t),
   have H2 : proj2 (cons_rep h t) = succ (len t),
     from refl (proj2 (cons_rep h t)),
   have H3 : proj2 (rep (cons h t)) = proj2 (cons_rep h t),
     from proj2_congr H1,
   trans H3 H2

theorem cons_ne_nil {A : (Type U)} (h : A) (t : list A) : cons h t ≠ nil
:= not_intro (assume R : cons h t = nil,
   have Heq1 : cons_rep h t = (nil_rep A),
      from abst_inj (list::inhab A) (cons_pred h t) (nil_pred A) R,
   have Heq2 : succ (len t) = zero,
      from proj2_congr Heq1,
   absurd Heq2 (succ_nz (len t)))

theorem flip_iff {a b : Bool} (H : a ↔ b) : ¬ a ↔ ¬ b
:= subst (refl (¬ a)) H

theorem cons_inj {A : (Type U)} {h1 h2 : A} {t1 t2 : list A} : cons h1 t1 = cons h2 t2 → h1 = h2 ∧ t1 = t2
:= assume Heq : cons h1 t1 = cons h2 t2,
   have Heq_rep : (cons_rep h1 t1) = (cons_rep h2 t2),
     from abst_inj (list::inhab A) (cons_pred h1 t1) (cons_pred h2 t2) Heq,
   have Heq_len : len t1 = len t2,
     from succ_inj (proj2_congr Heq_rep),
   have Heq_fn  : (λ n, if n = len t2 then some h1 else (fn t1) n) =
                  (λ n, if n = len t2 then some h2 else (fn t2) n),
     from subst (proj1_congr Heq_rep) Heq_len,
   have Heq_some : some h1 = some h2,
     from calc some h1 = if len t2 = len t2 then some h1 else (fn t1) (len t2)  : by simp
                   ... = if len t2 = len t2 then some h2 else (fn t2) (len t2)  : congr1 Heq_fn (len t2)
                   ... = some h2                                                : by simp,
   have Heq_head : h1 = h2,
     from optional::injectivity Heq_some,
   have Heq_fn_t : fn t1 = fn t2,
     from funext (λ i,
                    or_elim (em (i = len t2))
                      (λ Heqi : i = len t2,
                         have Hlti2 : ¬ (i < len t2),
                           from eq_to_not_lt Heqi,
                         have Hlti1 : ¬ (i < len t1),
                           from subst Hlti2 (symm Heq_len),
                         have Ht1 : i < len t1 ↔ (fn t1) i ≠ (@none A),
                           from P_rep t1 i,
                         have Ht2 : i < len t2 ↔ (fn t2) i ≠ (@none A),
                           from P_rep t2 i,
                         have Heq1 : (fn t1) i = (@none A),
                           from not_not_elim ((flip_iff Ht1) ◂ Hlti1),
                         have Heq2 : (fn t2) i = (@none A),
                           from not_not_elim ((flip_iff Ht2) ◂ Hlti2),
                         trans Heq1 (symm Heq2))
                      (λ Hnei : i ≠ len t2,
                         calc fn t1 i = if i = len t2 then some h1 else (fn t1) i : by simp
                                  ... = if i = len t2 then some h2 else (fn t2) i : congr1 Heq_fn i
                                  ... = fn t2 i                                   : by simp)),
   have Heq_tail : t1 = t2,
     from ext Heq_len Heq_fn_t,
   and_intro Heq_head Heq_tail

theorem cons_inj_head {A : (Type U)} {h1 h2 : A} {t1 t2 : list A} : cons h1 t1 = cons h2 t2 → h1 = h2
:= assume H, and_eliml (cons_inj H)

theorem cons_inj_tail {A : (Type U)} {h1 h2 : A} {t1 t2 : list A} : cons h1 t1 = cons h2 t2 → t1 = t2
:= assume H, and_elimr (cons_inj H)

set_opaque list true
set_opaque cons true
set_opaque nil  true
set_opaque len  true
end
definition list := list::list