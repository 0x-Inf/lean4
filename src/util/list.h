/*
Copyright (c) 2013 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.

Author: Leonardo de Moura
*/
#pragma once
#include <iostream>
#include <iterator>
#include "rc.h"
#include "debug.h"

namespace lean {
/**
   \brief Basic list template.
*/
template<typename T>
class list {
    struct cell {
        MK_LEAN_RC()
        T      m_head;
        list   m_tail;
        cell(T const & h, list const & t):m_rc(1), m_head(h), m_tail(t) {}
        ~cell() {}
        void dealloc() { delete this; }
    };
    cell * m_ptr;
public:
    list():m_ptr(nullptr) {}
    list(T const & h, list const & t):m_ptr(new cell(h, t)) {}
    list(T const & h):m_ptr(new cell(h, list())) {}
    list(list const & s):m_ptr(s.m_ptr) { if (m_ptr) m_ptr->inc_ref(); }
    list(list&& s):m_ptr(s.m_ptr) { s.m_ptr = nullptr; }
    list(std::initializer_list<T> const & l):list() {
        auto it = l.end();
        while (it != l.begin()) {
            --it;
            *this = list(*it, *this);
        }
    }
    ~list() { if (m_ptr) m_ptr->dec_ref(); }

    list & operator=(list const & s) { LEAN_COPY_REF(list, s); }
    list & operator=(list && s) { LEAN_MOVE_REF(list, s); }

    operator bool() const { return m_ptr != nullptr; }

    friend bool is_nil(list const & l) { return l.m_ptr == nullptr; }
    friend T const & head(list const & l) { lean_assert(!is_nil(l)); return l.m_ptr->m_head; }
    friend list const & tail(list const & l) { lean_assert(!is_nil(l)); return l.m_ptr->m_tail; }

    friend bool is_eqp(list const & l1, list const & l2) { return l1.m_ptr == l2.m_ptr; }
    friend bool operator==(list const & l1, list const & l2) {
        cell * it1 = l1.m_ptr;
        cell * it2 = l2.m_ptr;
        while (it1 != nullptr && it2 != nullptr) {
            if (it1 == it2)
                return true;
            if (it1->m_head != it2->m_head)
                return false;
            it1 = it1->m_tail.m_ptr;
            it2 = it2->m_tail.m_ptr;
        }
        return it1 == nullptr && it2 == nullptr;
    }
    friend bool operator!=(list const & l1, list const & l2) { return !(l1 == l2); }

    /** \brief List iterator object. */
    class iterator {
        friend class list;
        cell const * m_it;
        iterator(cell const * it):m_it(it) {}
    public:
        typedef std::forward_iterator_tag iterator_category;
        typedef T         value_type;
        typedef unsigned  difference_type;
        typedef T const * pointer;
        typedef T const & reference;

        iterator(iterator const & s):m_it(s.m_it) {}
        iterator & operator++() { m_it = m_it->m_tail.m_ptr; return *this; }
        iterator operator++(int) { iterator tmp(*this); operator++(); return tmp; }
        bool operator==(iterator const & s) const { return m_it == s.m_it; }
        bool operator!=(iterator const & s) const { return !operator==(s); }
        T const & operator*() { lean_assert(m_it); return m_it->m_head; }
        T const * operator->() { lean_assert(m_it); return &(m_it->m_head); }
    };

    iterator begin() const { return iterator(m_ptr); }
    iterator end() const { return iterator(nullptr); }
};

template<typename T> inline list<T>         cons(T const & h, list<T> const & t) { return list<T>(h, t); }
template<typename T> inline T const &       car(list<T> const & l) { return head(l); }
template<typename T> inline list<T> const & cdr(list<T> const & l) { return tail(l); }

template<typename T> inline std::ostream & operator<<(std::ostream & out, list<T> const & l) {
    out << "(";
    bool first = true;
    list<T> const * ptr = &l;
    while (*ptr) {
        if (first)
            first = false;
        else
            out << " ";
        out << head(*ptr);
        ptr = &tail(*ptr);
    }
    out << ")";
    return out;
}

template<typename T> unsigned length(list<T> const & l) {
    unsigned r = 0;
    list<T> const * it = &l;
    while (*it) {
        r++;
        it = &tail(*it);
    }
    return r;
}

/** \brief Return a list containing the elements in the subrange <tt>[begin, end)</tt>. */
template<typename It, typename T = typename It::value_type> list<T> copy_to_list(It const & begin, It const & end) {
    list<T> r;
    auto it = end;
    while (it != begin) {
        --it;
        r = cons(*it, r);
    }
    return r;
}
}
