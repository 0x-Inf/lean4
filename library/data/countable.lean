/-
Copyright (c) 2015 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.

Module: data.countable
Author: Leonardo de Moura

Type class for countable types
-/
import data.fintype data.list data.sum data.nat data.subtype
open option list nat

structure countable [class] (A : Type) :=
(pickle : A → nat) (unpickle : nat → option A) (picklek : ∀ a, unpickle (pickle a) = some a)

open countable

definition countable_fintype [instance] {A : Type} [h₁ : fintype A] [h₂ : decidable_eq A] : countable A :=
countable.mk
  (λ a, find a (elements_of A))
  (λ n, nth (elements_of A) n)
  (λ a, find_nth (fintype.complete a))

definition countable_nat [instance] : countable nat :=
countable.mk (λ a, a) (λ n, some n) (λ a, rfl)

definition countable_option [instance] {A : Type} [h : countable A] : countable (option A) :=
countable.mk
  (λ o, match o with
        | some a := succ (pickle a)
        | none := 0
        end)
  (λ n, if n = 0 then some none else some (unpickle A (pred n)))
  (λ o,
    begin
    cases o with [a],
      begin esimp end,
      begin esimp, rewrite [if_neg !succ_ne_zero, pred_succ, countable.picklek] end
    end)

section sum
variables {A B : Type}
variables [h₁ : countable A] [h₂ : countable B]
include h₁ h₂

definition pickle_sum : sum A B → nat
| (sum.inl a) := 2 * pickle a
| (sum.inr b) := 2 * pickle b + 1

definition unpickle_sum (n : nat) : option (sum A B) :=
if n mod 2 = 0 then
   match unpickle A (n div 2) with
   | some a := some (sum.inl a)
   | none   := none
   end
else
   match unpickle B ((n - 1) div 2) with
   | some b := some (sum.inr b)
   | none   := none
   end

open decidable
theorem unpickle_pickle_sum : ∀ s : sum A B, unpickle_sum (pickle_sum s) = some s
| (sum.inl a) :=
  assert aux : 2 > 0, from dec_trivial,
  begin
    esimp [pickle_sum, unpickle_sum],
    rewrite [mul_mod_right, if_pos (eq.refl 0), mul_div_cancel_left _ aux, countable.picklek]
  end
| (sum.inr b) :=
  assert aux₁ : 2 > 0,       from dec_trivial,
  assert aux₂ : 1 mod 2 = 1, by rewrite [modulo_def],
  assert aux₃ : 1 ≠ 0,       from dec_trivial,
  begin
    esimp [pickle_sum, unpickle_sum],
    rewrite [add.comm, add_mul_mod_self_left aux₁, aux₂, if_neg aux₃, add_sub_cancel_left,
             mul_div_cancel_left _ aux₁, countable.picklek]
  end

definition countable_sum [instance] : countable (sum A B) :=
countable.mk
  (λ s, pickle_sum s)
  (λ n, unpickle_sum n)
  (λ s, unpickle_pickle_sum s)
end sum

section prod
variables {A B : Type}
variables [h₁ : countable A] [h₂ : countable B]
include h₁ h₂

definition pickle_prod : A × B → nat
| (a, b) := mkpair (pickle a) (pickle b)

definition unpickle_prod (n : nat) : option (A × B) :=
match unpair n with
| (n₁, n₂) :=
  match unpickle A n₁ with
  | some a :=
    match unpickle B n₂ with
    | some b := some (a, b)
    | none   := none
    end
  | none   := none
  end
end

theorem unpickle_pickle_prod : ∀ p : A × B, unpickle_prod (pickle_prod p) = some p
| (a, b) :=
  begin
    esimp [pickle_prod, unpickle_prod, prod.cases_on],
    rewrite [unpair_mkpair],
    esimp,
    rewrite [*countable.picklek]
  end

definition countable_product [instance] : countable (A × B) :=
countable.mk
  pickle_prod
  unpickle_prod
  unpickle_pickle_prod
end prod

section list
variables {A : Type}
variables [h : countable A]
include h

definition pickle_list_core : list A → nat
| []     := 0
| (a::l) := mkpair (pickle a) (pickle_list_core l)

theorem pickle_list_core_cons (a : A) (l : list A) : pickle_list_core (a::l) = mkpair (pickle a) (pickle_list_core l) :=
rfl

definition pickle_list (l : list A) : nat :=
mkpair (length l) (pickle_list_core l)

definition unpickle_list_core : nat → nat → option (list A)
| 0        v  := some []
| (succ n) v  :=
  match unpair v with
  | (v₁, v₂) :=
    match unpickle A v₁ with
    | some a :=
      match unpickle_list_core n v₂ with
      | some l := some (a::l)
      | none   := none
      end
    | none   := none
    end
  end

theorem unpickle_list_core_succ (n v : nat) :
  unpickle_list_core (succ n) v =
    match unpair v with
    | (v₁, v₂) :=
      match unpickle A v₁ with
      | some a :=
        match unpickle_list_core n v₂ with
        | some l := some (a::l)
        | none   := none
        end
      | none   := none
      end
    end
:= rfl

definition unpickle_list (n : nat) : option (list A) :=
match unpair n with
| (l, v) := unpickle_list_core l v
end

theorem unpickle_pickle_list_core : ∀ l : list A, unpickle_list_core (length l) (pickle_list_core l) = some l
| []     := rfl
| (a::l) :=
  begin
    rewrite [pickle_list_core_cons, length_cons, add_one (length l), unpickle_list_core_succ],
    rewrite [unpair_mkpair],
    esimp [prod.cases_on],
    rewrite [unpickle_pickle_list_core l],
    rewrite [countable.picklek],
  end

theorem unpickle_pickle_list (l : list A) : unpickle_list (pickle_list l) = some l :=
begin
  esimp [pickle_list, unpickle_list],
  rewrite [unpair_mkpair],
  esimp [prod.cases_on],
  apply unpickle_pickle_list_core
end

definition countable_list [instance] : countable (list A) :=
countable.mk
  pickle_list
  unpickle_list
  unpickle_pickle_list
end list

definition countable_of_left_injection
             {A B : Type} [h₁ : countable A]
             (f : B → A) (finv : A → option B) (linv : ∀ b, finv (f b) = some b) : countable B :=
countable.mk
  (λ b, pickle (f b))
  (λ n,
    match unpickle A n with
    | some a := finv a
    | none   := none
    end)
  (λ b,
    begin
      esimp,
      rewrite [countable.picklek],
      esimp [option.cases_on],
      rewrite [linv]
    end)

/-
Choice function for countable types  and decidable predicates.
We provide the following API

choose      {A : Type} {p : A → Prop} [c : countable A] [d : decidable_pred p] : (∃ x, p x) → A :=
choose_spec {A : Type} {p : A → Prop} [c : countable A] [d : decidable_pred p] (ex : ∃ x, p x) : p (choose ex) :=
-/
section find_a
parameters {A : Type} {p : A → Prop} [c : countable A] [d : decidable_pred p]
include c
include d

private definition pn (n : nat) : Prop :=
match unpickle A n with
| some a := p a
| none   := false
end

private definition decidable_pn : decidable_pred pn :=
λ n,
match unpickle A n with
| some a := λ e : unpickle A n = some a,
  match d a with
  | decidable.inl t :=
    begin
      unfold pn, rewrite e, esimp [option.cases_on],
      exact (decidable.inl t)
    end
  | decidable.inr f :=
    begin
      unfold pn, rewrite e, esimp [option.cases_on],
      exact (decidable.inr f)
    end
  end
| none   := λ e : unpickle A n = none,
  begin
    unfold pn, rewrite e, esimp [option.cases_on],
    exact decidable_false
  end
end (eq.refl (unpickle A n))

private definition ex_pn_of_ex : (∃ x, p x) → (∃ x, pn x) :=
assume ex,
obtain (w : A) (pw : p w), from ex,
exists.intro (pickle w)
  begin
    unfold pn, rewrite [picklek], esimp, exact pw
  end

private lemma unpickle_ne_none_of_pn {n : nat} : pn n → unpickle A n ≠ none :=
assume pnn e,
begin
  rewrite [▸ (match unpickle A n with | some a := p a | none   := false end) at pnn],
  rewrite [e at pnn], esimp [option.cases_on] at pnn,
  exact (false.elim pnn)
end

open subtype

private lemma of_nat (n : nat) : pn n → { a : A | p a } :=
match unpickle A n with
| some a := λ (e : unpickle A n = some a),
  begin
    unfold pn, rewrite e, esimp [option.cases_on], intro pa,
    exact (tag a pa)
  end
| none   := λ (e : unpickle A n = none) h, absurd e (unpickle_ne_none_of_pn h)
end (eq.refl (unpickle A n))

private definition find_a : (∃ x, p x) → {a : A | p a} :=
assume ex : ∃ x, p x,
have exn  : ∃ x, pn x, from ex_pn_of_ex ex,
let r     : nat := @nat.choose pn decidable_pn exn in
have pnr  : pn r, from @nat.choose_spec pn decidable_pn exn,
of_nat r pnr
end find_a

namespace countable
open subtype

definition choose {A : Type} {p : A → Prop} [c : countable A] [d : decidable_pred p] : (∃ x, p x) → A :=
assume ex, elt_of (find_a ex)

theorem choose_spec {A : Type} {p : A → Prop} [c : countable A] [d : decidable_pred p] (ex : ∃ x, p x) : p (choose ex) :=
has_property (find_a ex)

theorem axiom_of_choice {A : Type} {B : A → Type} {R : Π x, B x → Prop} [c : Π a, countable (B a)] [d : ∀ x y, decidable (R x y)]
                        : (∀x, ∃y, R x y) → ∃f, ∀x, R x (f x) :=
assume H,
have H₁ : ∀x, R x (choose (H x)), from take x, choose_spec (H x),
exists.intro _ H₁

theorem skolem {A : Type} {B : A → Type} {P : Π x, B x → Prop} [c : Π a, countable (B a)] [d : ∀ x y, decidable (P x y)]
               : (∀x, ∃y, P x y) ↔ ∃f, (∀x, P x (f x)) :=
iff.intro
  (assume H : (∀x, ∃y, P x y), axiom_of_choice H)
  (assume H : (∃f, (∀x, P x (f x))),
    take x, obtain (fw : ∀x, B x) (Hw : ∀x, P x (fw x)), from H,
      exists.intro (fw x) (Hw x))
end countable
