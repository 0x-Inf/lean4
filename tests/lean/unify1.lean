import data.list
open list

definition ID {A : Type} (a : A) := a

-- set_option trace.type_context.is_def_eq_detail true
-- set_option pp.purify_metavars false

#unify id (_ ++ (_ : list nat)), ([1, 2] ++ [(3:nat), 4])

#unify (_ ++ (_ : list nat)), id ([1, 2] ++ [(3:nat), 4])

#unify (_ ++ (_ : list nat)), ID ([1, 2] ++ [(3:nat), 4])

#unify (_ ++ (_ : list nat)), ID (ID ([1, 2] ++ [(3:nat), 4]))

#unify ID (_ ++ (_ : list nat)), ID ([1, 2] ++ [(3:nat), 4])

#unify ID (_ ++ (_ : list nat)), ([1, 2] ++ [(3:nat), 4])

-- The following one fails because both sides are productive, and
-- we need to unfolding steps in the lhs to get the same
-- head symbol in the rhs
-- #unify ID (ID (_ ++ (_ : list nat))), ([1, 2] ++ [(3:nat), 4])

#unify ([(1:nat)] ++ [2, 3, 4]), ([1, 2] ++ [(3:nat), 4])

constant y1 : nat
constant y2 : nat

#unify (y1 :: _) ++ _, (let l := [y1] in l ++ [y2])
