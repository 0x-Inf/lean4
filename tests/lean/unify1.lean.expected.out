id (?m_1 ++ ?m_2) =?= [1, 2] ++ [3, 4]
id ([1, 2] ++ [3, 4]) =?= [1, 2] ++ [3, 4]
success
?m_1 ++ ?m_2 =?= id ([1, 2] ++ [3, 4])
[1, 2] ++ [3, 4] =?= id ([1, 2] ++ [3, 4])
success
?m_1 ++ ?m_2 =?= ID ([1, 2] ++ [3, 4])
[1, 2] ++ [3, 4] =?= ID ([1, 2] ++ [3, 4])
success
?m_1 ++ ?m_2 =?= ID (ID ([1, 2] ++ [3, 4]))
[1, 2] ++ [3, 4] =?= ID (ID ([1, 2] ++ [3, 4]))
success
ID (?m_1 ++ ?m_2) =?= ID ([1, 2] ++ [3, 4])
ID ([1, 2] ++ [3, 4]) =?= ID ([1, 2] ++ [3, 4])
success
ID (?m_1 ++ ?m_2) =?= [1, 2] ++ [3, 4]
ID ([1, 2] ++ [3, 4]) =?= [1, 2] ++ [3, 4]
success
[1] ++ [2, 3, 4] =?= [1, 2] ++ [3, 4]
[1] ++ [2, 3, 4] =?= [1, 2] ++ [3, 4]
success
y1 :: ?m_1 ++ ?m_2 =?= let l := [y1] in l ++ [y2]
[y1] ++ [y2] =?= let l := [y1] in l ++ [y2]
success
